packages/kit/src/runtime/server/endpoint.js

```diff
@@ -64,3 +65,24 @@ export async function render_endpoint(event, mod, state) {
 		throw error;
 	}
 }
+
+/**
+ * @param {import('types').RequestEvent} event
+ */
+export function is_endpoint_request(event) {
+	const result =
+		// These only exist for +server
+		['PUT', 'PATCH', 'DELETE'].includes(event.request.method) ||
+		// GET has accept text/html for pages
+		(event.request.method === 'GET' &&
+			negotiate(event.request.headers.get('accept') ?? '*/*', ['application/json', 'text/html']) ===
+				'application/json') ||
+		// POST with FormData is for actions
+		(event.request.method === 'POST' &&
+			!(event.request.headers.get('content-type') ?? '')
+				.split(';')
+				.some((part) =>
+					['multipart/form-data', 'application/x-www-form-urlencoded'].includes(part.trim())
+				));
+	return result;
+}
```

Comment:
This function is quite hard to read — something like this might be simpler:

```suggestion
export function is_endpoint_request(event) {
	const { method } = event.request;

	if (method === 'PUT' || method === 'PATCH' || method === 'DELETE') {
		return true;
	}

	if (method === 'GET') {
		const accept = event.request.headers.get('accept') ?? '*/*';
		return negotiate(accept, ['application/json', 'text/html']) === 'application/json';
	}

	if (method === 'POST') {
		const type = (event.request.headers.get('content-type') ?? '').split(';')[0];
		return type !== 'multipart/form-data' && type !== 'application/x-www-form-urlencoded';
	}

	return false;
}
```

That said, I'm not 100% sure this is the right logic. It's totally possible that someone would want to post form data to a `+server.js` handler, for example. I think we should maybe make it simpler: for both `GET` and `POST`, if the request is for `text/html` use `+page.server.js`, otherwise use `+server.js`:

```js
return negotiate(accept, ['*', 'text/html']) !== 'text/html';
```

Reply:
Not sure if this has the desired outcome in case of a progressive enhancement case, the request isn't for `text/html` in this case. The safest way to check this is probably to check if the `+page.server.js` contains `actions`, and if so, then formdata requests are routed to that.

Reply:
I think the rule needs to be that if you have a `+server.js` file, you're handling all non-HTML requests yourself. If we make assumptions to the contrary we're getting in people's way — the people likely to use content negotiation are the same people who are likely to be upset if we decide to invoke an action rather than a handler on their behalf.

If people _do_ want to invoke the action from the handler, they could potentially do something like this (once we add `fetch` to `RequestEvent`) — it's not the prettiest code...

```js
// src/routes/foo/+server.js
import { actions } from './+page.server.js';

export function POST({ fetch, request }) {
  return fetch(request.url, {
    method: request.method,
    body: request.body,
    headers: {
      accept: 'text/html'
    }
  });
}
```

...but it's a hell of a lot nicer than the alternative where you need to invoke a handler from an action because SvelteKit made that choice on your behalf.

Ultimately, if you have `+page.server.js` and `+server.js` in the same directory, you are opting for more control (potentially at the cost of some convenience), and a rule that says 'you are now responsible for handling non-HTML requests' is simpler than 'you are now responsible for handling non-HTML requests, except in the case where it's a `POST` request whose `content-type` begins with either `multipart/form-data` or `application/x-www-form-urlencoded`.

Reply:
That makes sense. What if we build in some kind of escape hatch for this situation? Like "if you want to call the form action use the `x-sveltekit-formaction` header"?

Reply:
That's the sort of thing we could potentially add later if people asked for it. My hunch is that they won't